TUIActivityIndicator Revamp!

avaidyam opened the pull request:
The next control on my hitlist for revamping: `TUIActivityIndicatorView`! I've renamed it to `TUIActivityIndicator` to remove the explanatory `View` from its name, and made it match the OS X/iOS activity indicator with customizable animations.

avaidyam committed:
Revamped TUIActivityIndicator.

avaidyam committed:
New animations and frames.

avaidyam committed:
Indicator layer proxy and constraints update.

avaidyam committed:
Fixed hidesWhenStopped and the example code.

jspahrsummers commented:
:-1: This is impressive work, but inappropriate for TwUI proper. `TUIActivityIndicatorView` has near parity with `UIActivityIndicatorView` right now – any extension of the scope beyond that is more appropriate for a third-party class.

avaidyam commented:
I'd like to mention that `TUIActivityIndicatorView` looks and acts nothing like `UIActivityIndicatorView`...

jspahrsummers commented:
The API has near parity, which is what I was referring to. If the appearance is different, that may have been a deliberate design choice. In any case, appearance changes are something we can discuss in issues or PRs – just because it doesn't look the same doesn't mean it should be replaced with a completely new API.

jspahrsummers commented:
Since it seems like you have lots of ideas like this, why not start a project for reusable TwUI views and controls? You could put these sorts of things into a library like that without bumping up against the scope of the TwUI framework itself.

jwilling commented:
One thing to note is that this might've been easier to implement if you had used a `CAReplicatorLayer` for the gears. :smiley:

avaidyam commented:
But it was already easy to implement? I'll look into `CAReplicatorLayer` though, thanks!

avaidyam commented:
As it is right now, I don't think TwUI does much. It has a half baked text engine, a powerful layer-backed view system, but nothing that works out of the box. With AppKit, it takes just a few lines to get a good app design working. With TwUI it takes MANY subclasses, and a lot of research and bug-fixing.  

I do understand that you don't want bloat added, but quoting the original "unveiling" post by Twitter, "And, since TwUI isn’t bound by the constraints of an existing API, developers can experiment with new features like block-based drawRect and layout.  …  We’ll continue to develop additional features and make improvements. And, we encourage you to experiment, as that will help us build a robust and exciting UI framework for the Mac."

And that's exactly what I'm doing, experimenting and developing new features for TwUI. I'd feel like as long as it's not something absolutely insane (like, for example, remote `CALayer` rendering and event handling for VNC-like uses), it should be fine to add in.

Back to the scope of this PR, quoting the `README`, "2. It's something so basic or important that TwUI really should have it." A `TUIActivityIndicatorView` is something I'd consider quite basic, and _preferably_ one that works as advertised. Following the first quote, the additional customization features should be perfectly okay too. Looking at the code, all I did to add these extra features was expose methods that would have been otherwise private, and add a set of animations and an indicator frame to _restore_ the old `TUIActivityIndicatorView` look. It's a simple hot swap of sorts. That's why I believe this PR should be okay to merge.

Sorry if I've come across as rude or raging, I'm just quite frustrated that the work I'm putting into a public framework repository is being turned down for being "too much". I'd much rather share these additions in the main repository so it becomes easier to use with a single download, out of the box, than add it to another repository as a subclass.

joshaber commented:
Our philosophy is that TwUI should be a foundation for UI. It shouldn't provide everything under the sun—just the core that anyone would need to make an app.

I'd say this situation is very similar to one in UIKit. You can find probably 100s of open source, customized UIKit activity indicators on the internet. But none of them belong in UIKit proper, even if UIKit were open source. `UIActivityIndicator` solves the general case simply.

It's impossible to overestimate the value of simplicity in an API. When it's a choice between simple API and complex API, we'll probably always side with simplicity.

Like @jspahrsummers, this is cool and you should certainly put it somewhere. But TwUI isn't that place.

To put it another way, we'd rather TwUI were a core around which people built cool extensions than a monolithic, all-encompassing solution.

avaidyam commented:
Thank you for that, it really helps me understand what TwUI _is to become_. On that note, would this PR be up for consideration again, if the excess features were removed? At its minimum, I wanted the activity indicator to look and act like iOS and OS X. It was when I realized that the animations could be replaced with a simple array, and the drawing replaced with a drawRect block to the inner knob, that I added these properties. To be honest, all I did was factor out the usual mechanism into public methods, which is why I'm heavily contesting the closing of this PR. I can, and _will_ make the API simpler, if that would let this PR go through.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+// as it rotates as well, while using the basic .layer property allows you
+// to modify stationary layer properties - those that are not affected by
+// the animations on the indicator.
+@property (nonatomic, readonly) CALayer *layerProxy;
```
The layer proxy is just a one-line accessor to the internal `indicator.layer`. It just lets you do fancy things.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+@property (nonatomic, readonly, getter = isAnimating) BOOL animating;
+
+// Changes the speed of the indicator animation. Defaults to 1.0 seconds.
+@property (nonatomic, assign) CGFloat animationSpeed;
```
The animationSpeed is another single-line property that lets you speed up or down the animations.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+// mutable nature of the animations array, you may combine several
+// unrelated animations to achieve a single complex animation. Also note
+// that anything that is not a CAAnimation object will be discarded.
+@property (nonatomic, strong) NSMutableArray *animations;
```
As I was factoring out the animation to become more iOS/OS X-like, I realized that it was already an array, so why not make this array public?

jwilling replied:
I don't think making a mutable array public is a good idea. Since you're only accepting `CAAnimations`, why not just add a method that adds a single `CAAnimation` object onto the private array of animations?

avaidyam replied:
I could do that….. :D

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+// and a combination of the indicatorFrame and the animations array,
+// a completely custom look can be applied. This will only apply if
+// the activityIndicatorStyle is set to TUIActivityIndicatorStyleCustom.
+@property (nonatomic, copy) TUIViewDrawRect indicatorFrame;
```
A simple 3 line setter/getter to the internal `indicator.drawRect`.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+@end
+
+// Returns an indicator frame with a basic centered gray circle.
+extern TUIViewDrawRect TUIActivityIndicatorCircleFrame();
```
Functions the same way as `TUITextView`'s `drawFrame` property, and I thought it would be okay, since the idea already existed in TwUI.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+// Restarts the animation of the indicator quietly, without triggering an
+// animation start or stop. This method is best used when drawing or
+// animation code is updated.
+- (void)refreshAnimations;
```
Again, as I was factoring out methods, I decided to make one or two public.

avaidyam commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,118 @@
+// Returns an array of animations designed to pulse the indicator. Consists
+// of a scaling transformation paired with an opacity blend for pulsing.
+// Pass a CGFloat for the peak opacity of the pulse animation.
+extern NSArray * TUIActivityIndicatorPulseAnimations(CGFloat);
```
Similar to above, but instead of pre-packaged `drawFrames`, it's a pre-packaged animation array. (It's usually a single animation, too).

avaidyam commented on lib/UIKit/TUIView+Layout.h:
```diff
@@ -25,4 +25,6 @@
 
 - (TUIView *)relativeViewForName:(NSString *)name;
 
+- (void)setNeedsUpdateConstraints;
```
Mirrors `UIView` and `NSView`. Now the `TUILayoutManager` can go fully private. Another PR for that later, though.

jspahrsummers replied:
This shouldn't be in this pull request.

jspahrsummers replied:
^

jspahrsummers replied:
^

avaidyam commented:
@jspahrsummers @jwilling @joshaber I've highlighted the changes (which I think are minuscule) I've made to the `TUIActivityIndicator`. I can remove any of these, if you'd like.

jspahrsummers commented:
As per @joshaber and my comments, this will only be considered if the API matches `UIActivityIndicatorView`. We're just talking in circles now.

avaidyam committed:
Shuffled things around, to begin removing them.

avaidyam committed:
Updated indicator API to make it clear that they're accessors to the inner indicator view.

I swear I can count the number of lines of code that this stuff takes
up….

avaidyam committed:
Shift stuff around.

avaidyam commented:
@jspahrsummers I've made the API clear and refactored. The API _does_ match `UIActivityIndicatorView`, with the exception of `layerProxy` and `indicatorFrame` which are both literally one-line accessors into the underlying indicator view, which `UIActivityIndicatorView` explicitly hides from you.

@jwilling I've moved the animations array into 2-line methods. :)

jwilling commented:
@galaxas0 :flushed: That might be worse than it was before. If you want to keep the ability to add an array of animations, it's probably good just to make it a immutable array. Adding a mutable property of this type just opens doors for all sorts of open ends about behaviors adding animations after they've begun.

avaidyam commented:
I have provisions for that-- they call `-refreshAnimations` when added.

avaidyam commented:
Wait-- just kidding, I see what you mean. +_____+

avaidyam committed:
Keep it simple.

avaidyam commented:
I've instead reverted it to a mutable array, but made the note to call the refresh method. I've made provisions for any sort of refreshing issues already, as it's inherent.

avaidyam committed:
More visually appealing demo.

jwilling commented:
I just don't understand your reasoning for making the array readonly and mutable. If you don't want them to modify the array during animation, then make it readwrite and immutable. Better not presume your users will follow advice in a comment.

avaidyam commented:
Oh, you _can_ modify the array during animation. That's why I did that. Just add a call to `refreshAnimations`. The array won't be affected either way during animations because I've offloaded the animations into a `CAAnimationGroup`.

avaidyam committed:
Merge remote-tracking branch 'upstream/master' into upstream-activity-indicator

avaidyam committed:
Remove extraneous API features.

avaidyam commented:
@jspahrsummers I've removed the extraneous features. Is there anything else you'd like me to remove?

jspahrsummers commented:
I'll review this if you can pare down the interface to:

``` objc
typedef enum {
    TUIActivityIndicatorViewStyleWhite,
    TUIActivityIndicatorViewStyleGray
} TUIActivityIndicatorViewStyle;

@interface TUIActivityIndicatorView : TUIView

@property (nonatomic, assign) TUIActivityIndicatorViewStyle activityIndicatorViewStyle;
@property (nonatomic, assign) BOOL hidesWhenStopped;
@property (nonatomic, readonly, getter = isAnimating) BOOL animating;

- (id)initWithActivityIndicatorViewStyle:(TUIActivityIndicatorStyle)style;
- (void)startAnimating;
- (void)stopAnimating;

@end
```

Note the name as well. I think we should follow the naming established by `UIActivityIndicatorView`, rather than `NSProgressIndicator`, since the former is more recent.

avaidyam commented:
I was going with `TUIActivityIndicator` since the addition of `View` is redundant, because it's a `TUIView` subclass by default. And why not the `animationSpeed`?

jspahrsummers commented:
It's not automatically implied that something within TwUI is a subclass of `TUIView`. There are several subclasses of AppKit classes within the framework. And, again, I'm talking about UIKit parity on this point.

> And why not the animationSpeed?

There should be no reason to customize the speed of the animation if the animation itself is not customizable.

avaidyam commented:
But I did make the animation customizable? Would it be alright to remove this `animationSpeed` and let animations have their own times?

jspahrsummers commented:
The interface I proposed above intentionally does not include customizable animations. The feature is unnecessary and difficult to use (not to mention maintain).

Another thing to keep in mind: TwUI is open source. If someone doesn't like how a component works, or wants more customization, they can always build their own based on the existing source code. We don't need to provide the kitchen sink.

avaidyam committed:
Rename and reorder stuff.

avaidyam commented:
I've stripped the interface down a fair amount, and I am ready to remove more, but I'm really concerned that there's no way to add custom animations anymore. Perhaps two methods like `beginUpdates` and `endUpdates` for `UITableView` would work for this. I'm very _very_ keen on the animations being easy to modify, but I completely understand that just putting it out there is _not_ what TwUI wants to do. Would you be open to any other methods of "inserting" animations?

jspahrsummers commented:
Custom animations do not have to be part of `TUIActivityIndicatorView`. The class is meant to be very basic.

If someone wants a different animation, it's trivial to write a new class to do it. Built-in TwUI classes have no advantage over what a user of the framework could write themselves, so we can't use "it's hard to do" as a reason to put more stuff into TwUI.

avaidyam commented:
@jspahrsummers I understand, and I've removed the customization features. The header is now as simple as you've described.

avaidyam commented:
@jspahrsummers :mailbox_with_mail:

jspahrsummers commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,61 @@
+- (id)initWithActivityIndicatorStyle:(TUIActivityIndicatorStyle)style;
+
+// Initializes the activity indicator with both style and frame.
+- (id)initWithFrame:(CGRect)frame andActivityIndicatorStyle:(TUIActivityIndicatorStyle)style;
```
Parameter names should never include `and` in Cocoa.

jspahrsummers commented on lib/UIKit/TUIActivityIndicator.m:
```diff
@@ -0,0 +1,169 @@
+static CGFloat const TUIActivityIndicatorDefaultToothCount = 12.0f;
+static CGFloat const TUIActivityIndicatorDefaultToothWidth = 2.0f;
+
+static TUIViewDrawRect TUIActivityIndicatorGearFrame(CGFloat toothCount, CGFloat toothWidth, NSColor *toothColor) {
```
Why is this still here?

jspahrsummers commented on lib/UIKit/TUIActivityIndicator.m:
```diff
@@ -0,0 +1,169 @@
+	} copy];
+}
+
+static NSArray * TUIActivityIndicatorGearAnimations(CGFloat frameCount) {
```
Why is this still here?

jspahrsummers commented on lib/UIKit/TUIActivityIndicator.h:
```diff
@@ -0,0 +1,61 @@
+	TUIActivityIndicatorStyleGray
+} TUIActivityIndicatorStyle;
+
+@interface TUIActivityIndicator : TUIView
```
This class still needs to be renamed to `TUIActivityIndicatorView`.

jspahrsummers commented:
This code has changed very little – you've only modified the public interface to hide the features that we declined to add.

Please reopen this once you've _actually_ reduced the scope.

avaidyam commented:
Sure, I'll fix that.

avaidyam commented:
@jspahrsummers I've reduced the scope and removed the reusability blocks. If there's anything I missed, I can remove that as well.

avaidyam commented:
@jspahrsummers Is there anything else wrong with this that I should change? It's been 4 days since I've updated it.

jspahrsummers commented:
Please reopen this pull request if you think it's ready for review.

avaidyam commented:
I'm not given the option to re-open it on github.com as there's only a "Comment" button below.

jwilling commented:
Odd. I'll open it for you since you think it's ready.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+#import "TUILayoutConstraint.h"
+#import "TUICGAdditions.h"
+
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect){0, 0, 32, 32};
```
Initializing structs based on member order is fragile. Please use `CGRectMake` or C's designated initializer syntax.

avaidyam replied:
Actually, LLVM 4.2 provides a rather nifty way to do this by defining the struct member like so:

```
(CGRect) {
    .origin.x = 0, 
    .origin.y = 0,
    .size.width = 32, 
    .size.height = 32
};
```

jwilling replied:
That's not new in LLVM. :wink: Was introduced in C99 if I recall correctly.

avaidyam replied:
Wow, really? I read it in one of the Clang release notes.... or am I confusing things... +____+

jwilling replied:
And as a matter of fact, I believe that could be shortened to the following: 

```
CGRect rect = (CGRect) {
    .size.width = 32, 
    .size.height = 32
};
```

as the origin will be initialized to zero point automatically if not provided. Anyway, it's irrelevant. :stuck_out_tongue_winking_eye:

avaidyam replied:
I guess it's better to be explicit? ;D :dart:

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+#import "TUICGAdditions.h"
+
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect){0, 0, 32, 32};
+static TUIActivityIndicatorStyle TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
```
This should be `const`.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+
+- (void)setHidesWhenStopped:(BOOL)hide {
+	_hidesWhenStopped = hide;
+	if(!self.animating && !hide && self.proxyIndicator.hidden)
```
Checking `self.proxyIndicator.hidden` is unnecessary here.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+- (void)setActivityIndicatorStyle:(TUIActivityIndicatorStyle)style {
+	_activityIndicatorStyle = style;
+	
+	self.proxyIndicator.drawRect = ^(TUIView *indicator, CGRect rect) {
```
For clarity, this block should be set on initialization (not when the style is changed), and can read from `weakSelf.activityIndicatorStyle` instead of closing over the argument.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+		
+		CGContextRef ctx = TUIGraphicsGetCurrentContext();
+		CGContextTranslateCTM(ctx, radius, radius);
+		CGContextClearRect(ctx, rect);
```
This is done by `clearsContextBeforeDrawing`, and should be unnecessary here.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+- (void)startAnimating {
+	if(!self.animating) {
+		self.proxyIndicator.hidden = NO;
+		_animating = YES;
```
This should use a setter. Avoid setting ivars directly whenever possible.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,143 @@
+		CAKeyframeAnimation *rotate = [CAKeyframeAnimation animationWithKeyPath:@"transform.rotation.z"];
+		rotate.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
+		rotate.calculationMode = kCAAnimationDiscrete;
+		rotate.repeatCount = INT_MAX;
```
This should be `HUGE_VALF`, per the [CAMediaTiming documentation](https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CAMediaTiming_protocol/Introduction/Introduction.html#//apple_ref/doc/uid/TP40004511-CH1-SW5).

avaidyam replied:
Nice catch-- didn't know that. Time to read up on that!

jspahrsummers commented:
:mailbox_with_mail:

avaidyam committed:
Style fixes and fixed animation direction.

avaidyam commented:
@jspahrsummers :mailbox:

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,149 @@
+#import "TUICGAdditions.h"
+
+static TUIActivityIndicatorStyle const TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect) {
```
`(CGRect)` is unnecessary here. The type should be inferred from the variable being assigned into.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,149 @@
+static TUIActivityIndicatorStyle const TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect) {
+	.origin.x = 0,
+	.origin.y = 0,
```
These can be condensed into `.origin = CGPointZero`, or omitted entirely (as noted by @jwilling).

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,149 @@
+
+@interface TUIActivityIndicatorView ()
+
+@property (nonatomic, strong) TUIView *proxyIndicator;
```
This should be `readonly`, as it is only ever set once.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.m:
```diff
@@ -15,80 +15,149 @@
-		spinner.alpha = 0.2;
-		spinner.layer.cornerRadius = 10.0;
-		[self addSubview:spinner];
+		__block TUIActivityIndicatorView *weakSelf = self;
```
`__block` does not make a variable weak under ARC. This should be:

``` objc
__unsafe_unretained TUIActivityIndicatorView *weakSelf = self;
```

jspahrsummers commented:
This makes the activity indicator animate in the wrong direction. It animates counterclockwise (making it look like progress is being rewound), when it should animate clockwise.

avaidyam commented:
@jspahrsummers I've made some style fixes and it now animates properly. :mailbox_with_mail:

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.h:
```diff
@@ -17,23 +17,48 @@
+	// The standard gray style of indicator.
+	TUIActivityIndicatorStyleGray,
+	
+	// The standard gray style of indicator.
```
Incorrect comment here.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.h:
```diff
@@ -17,23 +17,48 @@
 
-- (id)initWithActivityIndicatorStyle:(TUIActivityIndicatorViewStyle)style;
+// The basic appearance of the activity indicator.
+// The default value is TUIActivityIndicatorStyleWhite.
```
We should use the gray style as the default, like it was before.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.h:
```diff
@@ -17,23 +17,48 @@
-} TUIActivityIndicatorViewStyle;
+	
+	// The standard white style of indicator.
+	TUIActivityIndicatorStyleWhite,
```
The values in this enum should be prefixed with `TUIActivityIndicatorViewStyle`.

jspahrsummers commented on lib/UIKit/TUIActivityIndicatorView.h:
```diff
@@ -17,23 +17,48 @@
+	
+	// The standard black style of indicator.
+	TUIActivityIndicatorStyleBlack
+} TUIActivityIndicatorStyle;
```
This enum should be `TUIActivityIndicatorViewStyle`.

jspahrsummers commented:
Added some more notes. Also commented again on the `setNeedsUpdateConstraints` comment that you've missed a few times now.

avaidyam commented:
@jspahrsummers Ah, my bad. It should be _really_ fixed now... college apps drive me nuts.

avaidyam commented:
@jspahrsummers :mailbox_with_mail:

jspahrsummers commented:
Thanks!

avaidyam commented:
At last! :D

jwilling commented:
I'm not altogether sure this is without bugs. This was taken from the scrolling example:

![](http://i.imgur.com/zMP4a.png)

avaidyam commented:
Whoah, uhh, that's weird and totally not expected. That didn't happen before either…. I'll put in a PR to fix this weirdness.

avaidyam commented:
Wait, this is absolutely confusing... how is this happening.....????
